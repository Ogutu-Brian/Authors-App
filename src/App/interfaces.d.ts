// interfaces
import { Articles } from "../store/actions/articles/interfaces";

export interface AppProps {
  deleteArticle: (articleId: number) => Function;
  getArticles: Function;
  articles: Articles;
  markArticleAsRead: (articleId: number) => Function;
  search: (text: string) => void;
}

export interface AppState {
  displayTable: boolean;
  articles: Articles;
}

export type VoidFunction = () => void;
