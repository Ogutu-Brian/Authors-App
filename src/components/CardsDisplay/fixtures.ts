export const userIconUrl = 'https://res.cloudinary.com/authorsheaven/image/upload/v1568827956/samples/Images/avatar_atlbnx.png';
