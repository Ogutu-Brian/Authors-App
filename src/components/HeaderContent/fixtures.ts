export const inactiveCardUrl = 'https://res.cloudinary.com/authorsheaven/image/upload/v1568836047/samples/Images/card_vyumen.svg';

export const activeCardUrl = 'https://res.cloudinary.com/authorsheaven/image/upload/v1568836047/samples/Images/card_active_lddmcj.svg';

export const activeTableUrl = 'https://res.cloudinary.com/authorsheaven/image/upload/v1568836317/samples/Images/table_active_lwycko.svg';

export const inactiveTableUrl = 'https://res.cloudinary.com/authorsheaven/image/upload/v1568836317/samples/Images/table_bgsjim.svg';
