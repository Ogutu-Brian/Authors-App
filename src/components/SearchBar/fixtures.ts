export const searchIcon = 'https://res.cloudinary.com/authorsheaven/image/upload/v1568827958/samples/Images/search_icon_yz5rn8.png';

export const userIcon = 'https://res.cloudinary.com/authorsheaven/image/upload/v1568827957/samples/Images/Oval_2_ccmkeh.png';
