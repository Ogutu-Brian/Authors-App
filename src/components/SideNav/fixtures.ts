export const logoIcon = 'https://res.cloudinary.com/authorsheaven/image/upload/v1568827957/samples/Images/logo_bcjhah.png';

export const dashBoardIcon = 'https://res.cloudinary.com/authorsheaven/image/upload/v1568827955/samples/Images/dashboard_icon_yy3ym4.svg';

export const sessionIcon = 'https://res.cloudinary.com/authorsheaven/image/upload/v1568827958/samples/Images/sessions_icon_nzp7bl.svg';

export const meetingsIcon = 'https://res.cloudinary.com/authorsheaven/image/upload/v1568827957/samples/Images/meetings_icon_re7t5p.svg';

export const filesIcon = 'https://res.cloudinary.com/authorsheaven/image/upload/v1568827956/samples/Images/files_board_icon_ctpdww.svg';

export const conversationsIcon = 'https://res.cloudinary.com/authorsheaven/image/upload/v1568827955/samples/Images/conversations_icon_tln8nx.svg';

export const settingsIcon = 'https://res.cloudinary.com/authorsheaven/image/upload/v1568827958/samples/Images/settings_icon_pdnabn.svg';
