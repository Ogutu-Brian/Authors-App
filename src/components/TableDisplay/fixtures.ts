export const deleteBtn = 'https://res.cloudinary.com/authorsheaven/image/upload/v1568827955/samples/Images/delete_vlcpos.svg';
