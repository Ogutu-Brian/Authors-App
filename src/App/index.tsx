// react libraries
import * as React from 'react';

// third party libraries
import { connect } from 'react-redux';

// styles
import './App.scss';

// components
import SideNav from '../components/SideNav';
import SearchBar from '../components/SearchBar';
import HeaderContent from '../components/HeaderContent';
import DayHeader from '../components/DayHeader';
import ArticlesView from '../views/ArticlesView';
import ArticleCardsView from '../views/ArticleCardsView';

// thunks
import { getArticles, deleteArticle, markAsRead, fullTextSearch } from '../store/actions/articles';

// interfaces
import { AppProps, AppState, VoidFunction } from './interfaces';
import { Article } from '../store/actions/articles/interfaces';
import { Event } from '../views/ArticleCardsView/interfaces';

class App extends React.Component<AppProps, AppState> {
  state = {
    displayTable: true,
    articles: [],
  };

  componentDidMount() {
    this.setExistingArticlesToState();
  }

  static getDerivedStateFromProps(props: any, state: AppState) {
    if (props.articles !== state.articles) {
      return {
        ...state,
        articles: props.articles
      }
    }

    return state;
  }

  /**
   * @description sets existing articles to state during mounting
   * 
   * @returns {void}
   */
  setExistingArticlesToState = (): void => {
    this.props.getArticles().then(() => {
      const { articles } = this.props;

      this.setState({
        articles
      });
    });
  }

  /**
   * @description marks an article as read
   * 
   * @param {string} articleId
   *
   * @returns {VoidFunction}
   */
  markAsRead = (articleId: number): VoidFunction => async () => {
    await this.props.markArticleAsRead(articleId);
  }

  /**
   * @description marks all articles as read
   * 
   * @returns {VoidFunction}
   */
  markAllAsRead = (): VoidFunction => async () => {
    const { articles } = this.state;

    await articles.forEach((article: Article) => {
      this.markAsRead(article.id)();
    });
  }

  /**
   * @description Helps in switching between grid view and tabular view
   * 
   * @param {boolean} isTable
   * 
   * @returns {VoidFunction}
   */
  switchView = (isTable: boolean): VoidFunction => (): void => {
    if (isTable) {
      this.setState({
        displayTable: true
      });
    } else {
      this.setState({
        displayTable: false
      });
    }
  }

  /**
   * @description Deletes an article
   *
   * @param {number} articleId
   *
   * @returns {VoidFunction}
   */
  deleteArticle = (articleId: number): VoidFunction => (): void => {
    this.props.deleteArticle(articleId);
  }

  /**
   * @description Handles full text search
   * 
   * @param {Event} event
   *
   * @returns {void}
   */
  fullTextSearch = (event: Event): void => {
    const text = event.target.value;
    this.props.search(text);
  }

  render() {
    const { displayTable, articles } = this.state;

    return (
      <div className="App">
        <SideNav />
        <div className="App__content">
          <SearchBar
            handleSearch={this.fullTextSearch}
          />
          <HeaderContent
            switchView={this.switchView}
            tableActive={displayTable}
          />
          <ArticleCardsView
            articles={articles}
            markAllAsRead={this.markAllAsRead}
          />
          <DayHeader />
          <ArticlesView
            articles={articles}
            displayTable={displayTable}
            deleteArticle={this.deleteArticle}
            markAsRead={this.markAsRead}
          />
        </div>
      </div>
    );
  }
}

const mapDispatchToProps = (dispatch: any) => ({
  deleteArticle: (articleId: number) => deleteArticle(dispatch, articleId),
  getArticles: () => getArticles(dispatch)(),
  markArticleAsRead: (articleId: number) => markAsRead(dispatch, articleId),
  search: (text: string) => fullTextSearch(dispatch, text),
});

const mapStateToProps = (state: any) => ({
  articles: state.articles.articles
});

export default connect(mapStateToProps, mapDispatchToProps)(App);
