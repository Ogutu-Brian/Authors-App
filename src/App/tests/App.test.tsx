// react libraries
import * as React from 'react';

// components
import App from '..';

// third party libraries
import { mount, configure } from 'enzyme';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { Provider } from 'react-redux';
import Adapter from 'enzyme-adapter-react-16';

// fixtures
import { articles } from '../fixtures';

configure({ adapter: new Adapter() });

const store = configureStore([thunk])({
  articles: {
    articles
  }
});

describe('Tests for App functionality', () => {
  const wrapper = mount(
    <Provider store={store}>
      <App />>
      </Provider>
  );
  const app = wrapper.find(App);

  it('should match app snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should swtich to cards option view and cards', () => {
    const cardsSwitch = app.find('.options__cards')

    cardsSwitch.simulate('click');
    expect(cardsSwitch.length).toEqual(1);
  });

  it('should switch to tabular view', () => {
    const tableSwitch = app.find('.options__table')

    tableSwitch.simulate('click');
    expect(tableSwitch.length).toEqual(1);
  });

  it('should mark an article as read', () => {
    const markAsReadButton = app.find('.read-btn__btn').first();

    markAsReadButton.simulate('click');
  });

  it('should delete an article', () => {
    const deleteBtn = app.find('.delete-btn__icon').first();

    deleteBtn.simulate('click');
  });

  it('should search for an article', () => {
    const searchBox = app.find('.search-box__input');
    searchBox.simulate('change', {
      target: {
        value: 'Sample text'
      }
    });
  });
});
