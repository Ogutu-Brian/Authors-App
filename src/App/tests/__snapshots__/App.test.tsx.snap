// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests for App functionality should match app snapshot 1`] = `ReactWrapper {}`;
